### PR TITLE
Feature/pi 2158 mas get unpaid work time

### PR DESCRIPTION
**Pull request by achimber-moj**

### CODE REVIEW

File: projects/manage-supervision-and-delius/src/main/kotlin/uk/gov/justice/digital/hmpps/service/SentenceService.kt [hunk 1]
```diff
@@ -62,20 +63,19 @@ class SentenceService(
...
         AdditionalSentence(length, amount, notes, type.description)
 
-    fun Person.toName() =
-        Name(forename, secondName, surname)
-
     fun Disposal.toOrder() =
         Order(description = type.description, length = length, startDate = date, endDate = expectedEndDate())
 
-    fun RequirementDetails.toRequirement(): Requirement {
+    fun RequirementDetails.toRequirement(eventId: Long, eventNumber: String): Requirement {
```
Comment (marcus-bcl) [inline projects/manage-supervision-and-delius/src/main/kotlin/uk/gov/justice/digital/hmpps/service/SentenceService.kt:78]: Do you still need to pass the eventId and eventNumber?

File: projects/manage-supervision-and-delius/src/main/kotlin/uk/gov/justice/digital/hmpps/service/SentenceService.kt [hunk 2]
```diff
@@ -116,6 +116,50 @@ class SentenceService(
...
+
+                else -> "${duration.toMinutesPart()} $minutesCompleted"
+            }
+
+            unpaidWorkHours ?: unpaidWorkMinutes ?: return "no time completed (of 1 hour)"
+            unpaidWorkHours ?: return "$unpaidWorkMinutes $totalMessage"
+            unpaidWorkMinutes ?: return "$unpaidWorkHours $completed $totalMessage"
+            return "$unpaidWorkHours and $unpaidWorkMinutes $totalMessage"
+        }
+    }
```
Comment (marcus-bcl) [inline projects/manage-supervision-and-delius/src/main/kotlin/uk/gov/justice/digital/hmpps/service/SentenceService.kt:161]: "(of 1 hour)" is hard-coded on line 156, also you might be able to simplify this with the Kotlin duration extension - e.g. something like:
```suggestion
    fun getUnpaidWorkTime(disposalId: Long): String {
        val hoursOrdered = requirementRepository.sumTotalUnpaidWorkHoursByDisposal(disposalId)
        val totalMessage = hoursOrdered
            ?.let { " (of $hoursOrdered hour${if (hoursOrdered != 1L) "s" else ""})"  } ?: ""

        val minutesCredited = upwAppointmentRepository.calculateUnpaidTimeWorked(eventId, eventNumber)
        val creditedMessage = Duration.ofMinutes(minutesCredited).toKotlinDuration()
            .toComponents { hours, minutes, _, _ -> 
                when {
                    hours == 0L -> "$minutes minute${if (minutes != 1) "s" else ""} completed"
                    minutes == 0 -> "$hours hour${if (hours != 1L) "s" else ""} completed"
                    else -> "$hours hour${if (hours != 1L) "s" else ""} $minutes minute${if (minutes != 1) "s" else ""} completed"
                }
            }
        return "$creditedMessage$totalMessage"
    }
```

Comment (marcus-bcl) [reply projects/manage-supervision-and-delius/src/main/kotlin/uk/gov/justice/digital/hmpps/service/SentenceService.kt:161]: You can play around with a few test cases in the playground here: https://pl.kotl.in/Don_1E9j7

File: projects/manage-supervision-and-delius/src/main/kotlin/uk/gov/justice/digital/hmpps/service/SentenceService.kt [hunk 3]
```diff
@@ -62,20 +63,19 @@ class SentenceService(
                 additionalSentences.map { it.toAdditionalSentence() }
             ),
             order = disposal?.toOrder(),
-            requirements = requirementRepository.getRequirements(id, eventNumber).map { it.toRequirement() },
-            courtDocuments = documentRepository.getCourtDocuments(id, eventNumber).map { it.toCourtDocument() }
+            requirements = requirementRepository.getRequirements(id, eventNumber)
+                .map { it.toRequirement(id, eventNumber) },
+            courtDocuments = documentRepository.getCourtDocuments(id, eventNumber).map { it.toCourtDocument() },
+            disposal?.id?.let { getUnpaidWorkTime(it) }
```
Comment (marcus-bcl) [inline projects/manage-supervision-and-delius/src/main/kotlin/uk/gov/justice/digital/hmpps/service/SentenceService.kt:69]: Do we also need to check whether they have any unpaid work requirements against the disposal here, and return null if not?

File: projects/manage-supervision-and-delius/src/main/kotlin/uk/gov/justice/digital/hmpps/integrations/delius/overview/entity/Requirement.kt [hunk 4]
```diff
@@ -157,6 +157,16 @@ interface RequirementRepository : JpaRepository<Requirement, Long> {
...
     fun getRequirements(id: Long, eventNumber: String): List<RequirementDetails>
+
+    @Query(
+        "SELECT SUM(r.length) " +
+            "FROM Requirement r " +
+            "JOIN  r.mainCategory mc " +
+            "JOIN  r.disposal " +
+            "WHERE r.disposal.id = :id " +
+            "AND mc.code = 'W'"
+    )
```
Comment (marcus-bcl) [inline projects/manage-supervision-and-delius/src/main/kotlin/uk/gov/justice/digital/hmpps/integrations/delius/overview/entity/Requirement.kt:168]: We should use multi-line strings for these queries:
```suggestion
    @Query("""
        SELECT SUM(r.length)
        FROM Requirement r
        JOIN  r.mainCategory mc
        JOIN  r.disposal
        WHERE r.disposal.id = :id
        AND mc.code = 'W'
    """)
```

**Review by marcus-bcl — COMMENTED** [review]
Looks good, few minor comments

File: projects/manage-supervision-and-delius/src/main/kotlin/uk/gov/justice/digital/hmpps/service/SentenceService.kt (hunk 1 above)
Comment (achimber-moj) [reply projects/manage-supervision-and-delius/src/main/kotlin/uk/gov/justice/digital/hmpps/service/SentenceService.kt]: updated and fields are no longer used

File: projects/manage-supervision-and-delius/src/main/kotlin/uk/gov/justice/digital/hmpps/integrations/delius/overview/entity/Requirement.kt (hunk 4 above)
Comment (achimber-moj) [reply projects/manage-supervision-and-delius/src/main/kotlin/uk/gov/justice/digital/hmpps/integrations/delius/overview/entity/Requirement.kt:168]: updated

File: projects/manage-supervision-and-delius/src/main/kotlin/uk/gov/justice/digital/hmpps/service/SentenceService.kt (hunk 2 above)
Comment (achimber-moj) [reply projects/manage-supervision-and-delius/src/main/kotlin/uk/gov/justice/digital/hmpps/service/SentenceService.kt:161]: updated

File: projects/manage-supervision-and-delius/src/main/kotlin/uk/gov/justice/digital/hmpps/service/SentenceService.kt (hunk 3 above)
Comment (achimber-moj) [reply projects/manage-supervision-and-delius/src/main/kotlin/uk/gov/justice/digital/hmpps/service/SentenceService.kt:70]: line 123 does this check and returns null for this scenario

File: projects/manage-supervision-and-delius/src/main/kotlin/uk/gov/justice/digital/hmpps/integrations/delius/sentence/entity/UpwAppointment.kt [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+interface UpwAppointmentRepository : JpaRepository<UpwAppointment, Long> {
+
+    @Query(
+        "SELECT SUM(u.minutesCredited) " +
+            "FROM UpwAppointment u " +
+            "JOIN  u.upwDetails upd " +
+            "JOIN  upd.disposal d " +
+            "WHERE d.id = :id " +
+            "AND u.softDeleted = 0 "
```
Comment (marcus-bcl) [inline projects/manage-supervision-and-delius/src/main/kotlin/uk/gov/justice/digital/hmpps/integrations/delius/sentence/entity/UpwAppointment.kt:50]: Minor thing - could you use a multi-line string here too?

**Review by marcus-bcl — APPROVED** [review]
Looks good

File: projects/manage-supervision-and-delius/src/main/kotlin/uk/gov/justice/digital/hmpps/integrations/delius/sentence/entity/UpwAppointment.kt (hunk 5 above)
Comment (achimber-moj) [reply projects/manage-supervision-and-delius/src/main/kotlin/uk/gov/justice/digital/hmpps/integrations/delius/sentence/entity/UpwAppointment.kt]: done

**Review by marcus-bcl — APPROVED** [review]

**Pull request merged**